# Python for Health, Economic, and Social Science
## Lab 1 — Instructor Solution & Expansion Notes

**Course:** Leverhulme Centre for Demographic Science (LCDS), Nuffield College Oxford  
**Lecturer:** Charles Rahal | **TA:** Jiani Yan  
**Day 1 theme:** Python object fundamentals — variables, collections, logic, and control flow

---

### How to use this file
- Each answer block is followed by an **Instructor Note** explaining *why* the solution works and what common student mistakes to watch for.
- **Expansion** sections go beyond the lab question, connecting to real demographic and social-science workflows.
- This file is for TA/instructor use. Distribute selectively or after the session.

---
## Part A: Strings and Numbers

### Q1: Intro String

In [ ]:
# Answer 1
name = "Jiani"
indicator = 'chronic diseases'
print(f"I am {name} and I want to study {indicator}.")

I am Alex and I want to study school_absence_rate.


> **Instructor note:** This immediately models the f-string syntax. Many beginners mix up `'` and `"` quotes or forget the `f` prefix — correct live if seen. The key teaching point is that variable names should be *descriptive*; `x` and `y` are never appropriate for social-science data.
>
> **Expansion — naming conventions in research code:**  
> In academic Python codebases you will see three conventions:  
> - `snake_case` — standard for variables and functions (`absence_rate`)  
> - `CamelCase` — standard for classes (`AreaRecord`)  
> - `UPPER_SNAKE` — constants (`MAX_INCOME_THRESHOLD`)  
> Demographic datasets (ONS, World Bank, EUROSTAT) typically use uppercase column names (`SP.DYN.LE00.IN`), so you will often need to rename them as soon as you load the data.

### Q2: Special Characters

In [2]:
# Answer 2
print("Area: Northside\n\tHealth indicator: clinic_visits\n\tSocial indicator: school_absence_rate\n\tEconomic indicator: median_income")

Area: Northside
	Health indicator: clinic_visits
	Social indicator: school_absence_rate
	Economic indicator: median_income



> **Expansion — multi-line strings for report generation:**
```python
# Triple-quoted strings are useful for embedding long report text:
report_template = """
Area: {area}
    Health indicator:   {health}
    Social indicator:   {social}
    Economic indicator: {economic}
"""
print(report_template.format(area="Northside", health="clinic_visits",
                              social="school_absence_rate", economic="median_income"))
```
> This pattern is the foundation of automated report generation — a core skill in public health surveillance and policy briefing workflows.

### Q3: Concatenation

In [3]:
# Answer 3
region = "Northside"
health = "clinic_visits"
social = "school_absence_rate"
economic = "median_income"

# Method 1: f-string (preferred)
print(f"{region} | {health} | {social} | {economic}")

# Method 2: join (useful when you have a dynamic list)
parts = [region, health, social, economic]
print(" | ".join(parts))

Northside | clinic_visits | school_absence_rate | median_income
Northside | clinic_visits | school_absence_rate | median_income



>
> **Expansion — `.join()` in demographic data pipelines:**
```python
# Typical use: building a pipe-delimited record for logging
area_record = {"area": "Northside", "clinic_visits": 120, "median_income": 29000}
log_line = " | ".join(f"{k}={v}" for k, v in area_record.items())
print(log_line)  # area=Northside | clinic_visits=120 | median_income=29000
```

### Q4: String Methods

In [4]:
# Answer 4
s1 = "clinic visits"
s2 = "SCHOOL ABSENCE RATE"
s3 = "median_income"

print(s1.title())                          # 1. Title case
print(s2.lower())                          # 2. Lower case
print(s3.replace("_", " ").title())        # 3. Replace underscore, then title case

Clinic Visits
school absence rate
Median Income



> **Expansion — building a column name normaliser:**
```python
raw_columns = [" Clinic_Visits ", "SCHOOL_ABSENCE", "Median Income", "  unemployment_rate"]
clean_columns = [col.strip().lower().replace(" ", "_") for col in raw_columns]
print(clean_columns)
# ['clinic_visits', 'school_absence', 'median_income', 'unemployment_rate']
```
> This pattern maps directly to `pandas` column renaming:
```python
import pandas as pd
df.columns = [c.strip().lower().replace(" ", "_") for c in df.columns]
```

### Q5: Integer Operations — five expressions equalling 144

In [5]:
# Answer 5
print(12 ** 2)         # exponentiation: 12 squared
print(150 - 6)         # subtraction
print(1440 // 10)      # integer (floor) division
print(288 % 200 + 56)  # modulo + addition
print(6 * 24)          # multiplication (hours in a cycle)

144
144
144
144
144



>
> **Expansion — operator precedence matters in research code:**
```python
# Without parentheses — wrong for an average!
wrong_avg = 120 + 95 + 140 / 3   # evaluates as 120 + 95 + 46.67 = 261.67
correct_avg = (120 + 95 + 140) / 3  # = 118.33
print(wrong_avg, correct_avg)
# Tip: in research code, always use parentheses around numerators and denominators.
```

### Q6: Floating Point and Formatting

In [7]:
# Answer 6
absent = 37
total = 512

absence_rate = absent / total                     # 1. decimal rate
print(f"{absence_rate * 100:.2f}%")               # 2. formatted percentage

7.23%



>
> **Expansion — full-precision vs. display precision:**
```python
rate = 37 / 512
print(rate)                # 0.072265625 — full precision, fine for computation
print(f"{rate:.2%}")       # 7.23% — uses % format spec directly (multiplies by 100)
print(f"{rate:.4f}")       # 0.0723 — four decimal places
# In pandas you can set global display options:
# pd.set_option('display.float_format', '{:.2f}'.format)
```
>
> **Common student mistake:** `round(absence_rate, 2)` gives `0.07`, not `7.23%`. Round *after* multiplying by 100 if you want a percentage rounded to 2 decimal places.

---
## Part B: Lists and Dictionaries

### Q7: Build a Dictionary

In [ ]:
# Answer 7
northside = {
    "area": "Northside",
    "clinic_visits": 120,
    "school_absence_rate": 0.07,
    "median_income": 29000
}

print(northside["area"])
print(northside["clinic_visits"])
print(northside["school_absence_rate"])
print(northside["median_income"])


>
> **Expansion — `dict.get()` for safe access:**
```python
# Accessing a missing key raises KeyError:
# northside["vaccination_rate"]  --> KeyError

# Safe access with a default:
vax = northside.get("vaccination_rate", "not recorded")
print(vax)  # not recorded
```
> `.get()` is essential when parsing variable-quality administrative datasets where not every record has every field.

### Q8: Update and Extend a Dictionary

In [ ]:
# Answer 8
central = {"area": "Central", "clinic_visits": 95, "median_income": 36000}

central["school_absence_rate"] = 0.05    # 1. add new key
central["clinic_visits"] = 102           # 2. update existing key
print(central)

> **Expansion — updating multiple fields at once with `.update()`:**
```python
# Useful when patching a record after data linkage:
updates = {"school_absence_rate": 0.055, "clinic_visits": 105, "gp_registrations": 4800}
central.update(updates)
print(central)
# Mimics a merge/patch operation — common in record linkage workflows.
```

### Q9: List Indexing and Slicing

In [ ]:
# Answer 9
indicators = ["clinic_visits", "school_absence_rate", "median_income",
              "unemployment_rate", "rent_burden"]

print(indicators[0])          # 1. first item
print(indicators[-1])         # 2. last item via negative index
print(indicators[2:5])        # 3. economic indicators (median_income, unemployment_rate, rent_burden)


> **Expansion — slicing strategies:**
```python
# Every other item:
print(indicators[::2])   # ['clinic_visits', 'median_income', 'rent_burden']

# Reverse the list:
print(indicators[::-1])

# All except the first:
print(indicators[1:])

# Last two:
print(indicators[-2:])
```
> These patterns carry directly into NumPy and pandas: `df.iloc[0:5]`, `arr[::2]`, `df.iloc[:, -1]`.

### Q10: Reordering by Index

In [ ]:
# Answer 10
wrong_order = ["Q3", "Q1", "Q4", "Q2"]
right_order = [wrong_order[1], wrong_order[3], wrong_order[0], wrong_order[2]]
print(right_order)

> **Expansion — sorting lists with `sorted()` and custom keys:**
```python
# For simple alphabetical sort:
print(sorted(wrong_order))           # ['Q1', 'Q2', 'Q3', 'Q4']

# Sorting a list of dicts by a field value (very common in data work):
areas = [{"name": "Riverside", "income": 25000},
         {"name": "Central",   "income": 36000},
         {"name": "Northside", "income": 29000}]

sorted_areas = sorted(areas, key=lambda d: d["income"], reverse=True)
for a in sorted_areas:
    print(a["name"], a["income"])
```

### Q11: Split a Record String

In [ ]:
# Answer 11
record = "Riverside|clinic_visits=140|school_absence_rate=0.09|median_income=25000"

parts = record.split("|")
print(parts)                         # 1. full list

area_name = parts[0]
print(area_name)                     # 2. area name

income_field = parts[3]              # "median_income=25000"
income_value = income_field.split("=")[1]
print(income_value)                  # 3. value string for median_income


>
> **Expansion — building a parser from a raw string record:**
```python
record = "Riverside|clinic_visits=140|school_absence_rate=0.09|median_income=25000"
parts = record.split("|")
parsed = {"area": parts[0]}
for field in parts[1:]:
    key, val = field.split("=")
    # Coerce numerics
    try:
        parsed[key] = int(val)
    except ValueError:
        parsed[key] = float(val)
print(parsed)
# {'area': 'Riverside', 'clinic_visits': 140, 'school_absence_rate': 0.09, 'median_income': 25000}
```
> This pattern is a simplified CSV parser. On Day 3/4 students will use `pandas.read_csv()` which does this automatically, but understanding what happens under the hood helps diagnose encoding and delimiter errors.

### Q12: Build a Snapshot from Parallel Lists

In [ ]:
# Answer 12
areas = ["Northside", "Central", "Riverside"]
clinic_visits = [120, 102, 140]
school_absence_rate = [0.07, 0.05, 0.09]
median_income = [29000, 36000, 25000]

# Index 1 = "Central"
central_snapshot = {
    "area": areas[1],
    "clinic_visits": clinic_visits[1],
    "school_absence_rate": school_absence_rate[1],
    "median_income": median_income[1]
}
print(central_snapshot)


> **Expansion — `zip()` to combine parallel lists cleanly:**
```python
# Building ALL snapshots at once:
all_snapshots = [
    {"area": a, "clinic_visits": cv, "absence": sar, "income": mi}
    for a, cv, sar, mi in zip(areas, clinic_visits, school_absence_rate, median_income)
]
for s in all_snapshots:
    print(s)
# This is exactly what pd.DataFrame() expects — a list of dicts.
```

---
## Part C: Derived Indicators and Data Checks

### Q13: Per-1,000 Rate

In [ ]:
# Answer 13
clinic_visits = [120, 102, 140]
population    = [2400, 3100, 2800]

for i in range(3):
    rate = clinic_visits[i] / population[i] * 1000
    print(f"Area {i+1}: {rate:.1f} visits per 1,000 residents")


> **Expansion — crude vs. age-standardised rates (conceptual bridge):**
```python
# Crude rate calculation:
events   = [120, 102, 140]
pop      = [2400, 3100, 2800]
crude_rates = [e/p * 1000 for e, p in zip(events, pop)]
print([round(r, 1) for r in crude_rates])  # [50.0, 32.9, 50.0]

# Age-standardisation (preview of what NumPy/pandas makes easy on Day 3-4):
# Standard population weights are applied to age-specific rates,
# removing confounding from different age structures.
# Example: ONS European Standard Population (ESP 2013).
```

### Q14: Reporting Sentence

In [8]:
central_snapshot = {
    "area": "Central", "clinic_visits": 102,
    "school_absence_rate": 0.05, "median_income": 36000
}


In [9]:
# Answer 14  (uses central_snapshot from Q12)

area   = central_snapshot["area"]
visits = central_snapshot["clinic_visits"]
absence_pct = central_snapshot["school_absence_rate"] * 100
income = central_snapshot["median_income"]

print(f"{area} recorded {visits} clinic visits, {absence_pct:.1f}% absence, and median income {income}.")

Central recorded 102 clinic visits, 5.0% absence, and median income 36000.


In [10]:
f"{absence_pct/100:.1%} absence"

'5.0% absence'

> **Expansion — templated reporting for multiple areas:**
```python
mini_data = [
    {"area": "Northside", "clinic_visits": 120, "school_absence_rate": 0.07, "median_income": 29000},
    {"area": "Central",   "clinic_visits": 102, "school_absence_rate": 0.05, "median_income": 36000},
    {"area": "Riverside", "clinic_visits": 140, "school_absence_rate": 0.09, "median_income": 25000},
]

for rec in mini_data:
    print(f"{rec['area']} recorded {rec['clinic_visits']} clinic visits, "
          f"{rec['school_absence_rate']*100:.1f}% absence, median income £{rec['median_income']:,}.")
```
> The `:,` format spec adds thousand-separators, producing `£29,000` — important for readable policy tables.

### Q15: Clean Messy Indicator Names

In [ ]:
# Answer 15
raw = [" Clinic_Visits ", "school_absence_rate", " MEDIAN_INCOME"]

cleaned = [item.strip().lower().replace("_", " ") for item in raw]
print(cleaned)


> **Expansion — defensive normalisation pipeline:**
```python
import re

def normalise_indicator_name(raw: str) -> str:
    """Lowercase, strip whitespace, collapse multiple spaces/underscores to single underscore."""
    cleaned = raw.strip().lower()
    cleaned = re.sub(r'[\s_]+', '_', cleaned)   # unify whitespace and underscores
    cleaned = re.sub(r'[^a-z0-9_]', '', cleaned) # remove any remaining non-alphanumeric
    return cleaned

examples = [" Clinic  Visits ", "school_Absence__Rate", "Median-Income(%)", " GP_Registrations "]
print([normalise_indicator_name(e) for e in examples])
```
> This is the kind of utility function you would put in a `utils.py` module in a real research project.

### Q16: Type Conversion Check

In [ ]:
# Answer 16
visits_str = "140"
absence_str = "0.09"
income_str = "25000"

visits_int  = int(visits_str)
absence_flt = float(absence_str)
income_int  = int(income_str)

print(visits_int,  type(visits_int))
print(absence_flt, type(absence_flt))
print(income_int,  type(income_int))

In [11]:
print(int(-2.1))

-2



> **Expansion — `int()` truncates; `round()` for rounding:**
```python
print(int(3.9))    # 3, NOT 4 — truncation, not rounding!
print(round(3.9))  # 4
print(int(-3.9))   # -3 — towards zero

# Safe cast with a fallback:
def safe_int(val, default=0):
    try:
        return int(float(val))  # handles '140.0' as well as '140'
    except (ValueError, TypeError):
        return default

print(safe_int("140"))    # 140
print(safe_int("n/a"))    # 0
print(safe_int(None))     # 0
```

### Q17: Build a Nested Dictionary

In [12]:
# Answer 17
area_profiles = {
    "Northside": {"clinic_visits": 120, "school_absence_rate": 0.07, "median_income": 29000},
    "Central":   {"clinic_visits": 102, "school_absence_rate": 0.05, "median_income": 36000},
    "Riverside": {"clinic_visits": 140, "school_absence_rate": 0.09, "median_income": 25000},
}

In [14]:
import pandas as pd
df = pd.DataFrame(area_profiles).T
df

,clinic_visits,school_absence_rate,median_income
Northside,120.0,0.07,29000.0
Central,102.0,0.05,36000.0
Riverside,140.0,0.09,25000.0



>
> **Expansion — converting a nested dict to a DataFrame:**
```python
import pandas as pd
df = pd.DataFrame(area_profiles).T   # .T = transpose so areas are rows
df.index.name = "area"
print(df)
```

### Q18: Query and Compare

In [15]:
# Answer 18  (uses area_profiles from Q17)
area_profiles = {
    "Northside": {"clinic_visits": 120, "school_absence_rate": 0.07, "median_income": 29000},
    "Central":   {"clinic_visits": 102, "school_absence_rate": 0.05, "median_income": 36000},
    "Riverside": {"clinic_visits": 140, "school_absence_rate": 0.09, "median_income": 25000},
}

# 1. Riverside school absence rate
print(area_profiles["Riverside"]["school_absence_rate"])

# 2. Income gap: Central minus Riverside
income_gap = area_profiles["Central"]["median_income"] - area_profiles["Riverside"]["median_income"]
print(f"Income gap (Central – Riverside): £{income_gap:,}")

# 3. Combined clinic visits: Northside + Central
combined_visits = area_profiles["Northside"]["clinic_visits"] + area_profiles["Central"]["clinic_visits"]
print(f"Combined clinic visits (Northside + Central): {combined_visits}")

0.09
Income gap (Central – Riverside): £11,000
Combined clinic visits (Northside + Central): 222


---
## Part D: Integrated Mini-Case

### Q19: Mini Case Setup

In [16]:
# Answer 19
mini_data = [
    {"area": "Northside", "clinic_visits": 120, "school_absence_rate": 0.07, "median_income": 29000},
    {"area": "Central",   "clinic_visits": 102, "school_absence_rate": 0.05, "median_income": 36000},
    {"area": "Riverside", "clinic_visits": 140, "school_absence_rate": 0.09, "median_income": 25000},
]

# 1. Full record for Central
print(mini_data[1])

# 2. Median income for Northside
print(mini_data[0]["median_income"])

# 3. Clinic visits for Riverside
print(mini_data[2]["clinic_visits"])

{'area': 'Central', 'clinic_visits': 102, 'school_absence_rate': 0.05, 'median_income': 36000}
29000
140



> **Expansion — looping over the list of dicts:**
```python
# Find the area with the highest clinic visits:
max_visits_area = max(mini_data, key=lambda d: d["clinic_visits"])
print(max_visits_area["area"], max_visits_area["clinic_visits"])

# Filter: areas with absence rate above 0.06:
high_absence = [d["area"] for d in mini_data if d["school_absence_rate"] > 0.06]
print(high_absence)
```

### Q20: One-Line Summary Builder

In [ ]:
# Answer 20
central = {"area": "Central", "clinic_visits": 102, "school_absence_rate": 0.05, "median_income": 36000}

parts = [
    central["area"],
    f"visits: {central['clinic_visits']}",
    f"absence: {central['school_absence_rate']*100:.1f}%",
    f"income: {central['median_income']}"
]

print(" | ".join(parts))

### Q21 (Stretch): Debugging and Refactor

In [ ]:
# Answer 21 — three bugs: wrong capitalisation, string+int, string not converted
area = "Riverside"          # was 'Area' (NameError on print because 'area' ≠ 'Area')
clinic_visits_str = 140     # was '140' — can't add int to str
extra_visits = 5

print(area + " visits next month: " + str(clinic_visits_str + extra_visits))


> **Expansion — reading a TypeError traceback:**
```
TypeError: can only concatenate str (not "int") to str
```
> Python tells you exactly which types are clashing. The fix is always: either wrap the int in `str()`, or use an f-string which handles conversion automatically:
```python
print(f"{area} visits next month: {clinic_visits_str + extra_visits}")
```
>
> **Common student mistake:** `print(area + " visits next month: " + clinic_visits_str + extra)` — students forget that even though they "see" numbers, Python sees types. Connecting this to the `type()` function from Q16 closes the loop.

---
*Solution notebook prepared for LCDS Oxford Python Course, March 2026.  
For instructor and TA use. Questions: teaching@demography.ox.ac.uk*